File: api/models/scope.py
"""
System Scope models for FastAPI
"""
from enum import Enum
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


class SystemType(str, Enum):
    """System types for scope definition"""
    SUBSYSTEM = "subsystem"
    API = "api"
    BACKEND = "backend"
    FULLSYSTEM = "fullsystem"
    EMBEDDED = "embedded"
    OTHER = "other"


class SystemScopeBase(BaseModel):
    """Base System Scope attributes"""
    name: str = Field(..., description="Name of the system being analyzed")
    system_type: SystemType = Field(..., description="Type of system being analyzed")
    description: Optional[str] = Field(None, description="Detailed description of the system scope")
    boundaries: Optional[List[str]] = Field(default_factory=list, description="System boundaries")
    objectives: Optional[List[str]] = Field(default_factory=list, description="Analysis objectives")
    stakeholders: Optional[List[str]] = Field(default_factory=list, description="Stakeholders involved")


class SystemScopeCreate(SystemScopeBase):
    """Used for creating a new system scope"""
    scope_id: Optional[str] = Field(None, description="Optional unique scope identifier")


class SystemScopeUpdate(BaseModel):
    """Used for updating an existing system scope"""
    name: Optional[str] = None
    system_type: Optional[SystemType] = None
    description: Optional[str] = None
    boundaries: Optional[List[str]] = None
    objectives: Optional[List[str]] = None
    stakeholders: Optional[List[str]] = None


class SystemScope(SystemScopeBase):
    """Full system scope model with ID and timestamps"""
    scope_id: str = Field(..., description="Unique scope identifier")
    created_at: datetime
    updated_at: datetime
    
    class Config:
        orm_mode = True


class SystemScopeList(BaseModel):
    """List of system scopes"""
    scopes: List[SystemScope]
    total: int

File: frontend/src/api/scope.ts
/**
 * System Scope API client
 */
import apiClient from './index';

/**
 * System type enumeration
 */
export enum SystemType {
  SUBSYSTEM = 'subsystem',
  API = 'api',
  BACKEND = 'backend',
  FULLSYSTEM = 'fullsystem',
  EMBEDDED = 'embedded',
  OTHER = 'other'
}

/**
 * SystemScope interface
 */
export interface SystemScope {
  scope_id: string;
  name: string;
  system_type: SystemType;
  description?: string;
  boundaries?: string[];
  objectives?: string[];
  stakeholders?: string[];
  created_at: string;
  updated_at: string;
}

/**
 * SystemScopeCreate interface - for creating new scopes
 */
export interface SystemScopeCreate {
  name: string;
  system_type: SystemType;
  description?: string;
  boundaries?: string[];
  objectives?: string[];
  stakeholders?: string[];
  scope_id?: string;
}

/**
 * SystemScopeUpdate interface - for updating existing scopes
 */
export interface SystemScopeUpdate {
  name?: string;
  system_type?: SystemType;
  description?: string;
  boundaries?: string[];
  objectives?: string[];
  stakeholders?: string[];
}

/**
 * SystemScopeList interface - for listing scopes
 */
export interface SystemScopeList {
  scopes: SystemScope[];
  total: number;
}

/**
 * System Scope API methods
 */
export const scopeApi = {
  /**
   * Get a list of all system scopes
   */
  async getAll(skip = 0, limit = 100): Promise<SystemScopeList> {
    try {
      const queryParams = `?skip=${skip}&limit=${limit}`;
      return await apiClient.get(`/scope${queryParams}`);
    } catch (error) {
      console.error('Error fetching scopes:', error);
      throw error;
    }
  },

  /**
   * Get a specific system scope by ID
   */
  async getById(scopeId: string): Promise<SystemScope> {
    try {
      return await apiClient.get(`/scope/${scopeId}`);
    } catch (error) {
      console.error(`Error fetching scope ${scopeId}:`, error);
      throw error;
    }
  },

  /**
   * Create a new system scope
   */
  async create(scope: SystemScopeCreate): Promise<SystemScope> {
    try {
      return await apiClient.post('/scope', scope);
    } catch (error) {
      console.error('Error creating scope:', error);
      throw error;
    }
  },

  /**
   * Update an existing system scope
   */
  async update(scopeId: string, scope: SystemScopeUpdate): Promise<SystemScope> {
    try {
      return await apiClient.put(`/scope/${scopeId}`, scope);
    } catch (error) {
      console.error(`Error updating scope ${scopeId}:`, error);
      throw error;
    }
  },

  /**
   * Delete a system scope
   */
  async delete(scopeId: string): Promise<void> {
    try {
      return await apiClient.delete(`/scope/${scopeId}`);
    } catch (error) {
      console.error(`Error deleting scope ${scopeId}:`, error);
      throw error;
    }
  }
};

export default scopeApi;
